### PR TITLE
Magiceden polygon nft trades

### PR DESCRIPTION
**Pull request by springzh**
Brief comments on the purpose of your changes:

This is for Magic Eden marketplace on Polygon.

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Review by 0xRobin — CHANGES_REQUESTED** [review]
Hi sir,
Mostly the same comments is in the other PRs.
- remove mints model
- add uniqueness tests
- improve prices/tokens sources setup

Overall really great additions to the polygon dataset if we can tackle the open concerns and maybe also add in some seed tests later to validate data correctness of some crucial columns with some sample transactions. 

And sorry if any of my comment/remarks are a bit short/bold, just trying to push these reviews through before the weekend so you can move forward with them! :)

**Comment by 0xBoxer** [comment]
Hey @springzh, thanks for tackling these!

I think we should be able to include the royalty payments into the magic eden model.

https://polygonscan.com/tx/0xc55e0fee57f076d1944f1e9d22932c9e89e37debaee70572c56c86aca969c1af

The order of: 
- first token transfer -> payment
- second token transfer -> marketplace fee
- third token transfer -> royalty payment is consistent between all trades

Please incorporate royalty and marketplace fees into this PR

File: models/magiceden/polygon/magiceden_polygon_events.sql [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+price_list AS (
+    SELECT contract_address,
+          minute,
+          price,
+          decimals,
+          symbol
+     FROM {{ source('prices', 'usd') }} p
+     WHERE blockchain = 'polygon'
+        AND contract_address IN ( SELECT DISTINCT currency_contract FROM all_events)
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_events.sql]: Updated.

File: models/magiceden/polygon/magiceden_polygon_events.sql [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  'magiceden' AS project,
+  'v1' AS version,
+  a.evt_tx_hash AS tx_hash,
+  date_trunc('day', a.evt_block_time) AS block_date,
+  a.evt_block_time AS block_time,
+  a.evt_block_number AS block_number,
+  amount_raw / power(10, p.decimals) * p.price AS amount_usd,
+  amount_raw / power(10, p.decimals) AS amount_original,
+  amount_raw,
+  CASE WHEN p.symbol = 'WMATIC' THEN 'MATIC' ELSE p.symbol END AS currency_symbol,
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_events.sql]: Updated.

File: models/magiceden/polygon/magiceden_polygon_events.sql [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  p.contract_address AS currency_contract,
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_events.sql]: Updated.

File: models/magiceden/polygon/magiceden_polygon_fees.sql [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+CAST(platform_fee_percentage AS DOUBLE) AS platform_fee_percentage,
+CAST(royalty_fee_amount_raw AS DOUBLE) AS royalty_fee_amount_raw,
+royalty_fee_amount,
+royalty_fee_amount_usd,
+CAST(royalty_fee_percentage AS DOUBLE) AS royalty_fee_percentage,
+royalty_fee_receive_address,
+royalty_fee_currency_symbol,
+token_standard,
+trade_type,
+CAST(number_of_items AS DECIMAL(38,0)) AS number_of_items,
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_fees.sql]: Updated.

File: models/magiceden/polygon/magiceden_polygon_mints.sql [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      aggregator_name,
+      aggregator_address,
+      block_number,
+      tx_hash,
+      tx_from,
+      tx_to,
+      unique_trade_id
+FROM {{ ref('magiceden_polygon_events') }}
+WHERE evt_type = 'Mint'
+;
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_mints.sql]: Updated.

File: models/magiceden/polygon/magiceden_polygon_schema.yml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  - name: magiceden_polygon_events
+    meta:
+      blockchain: polygon
+      project: magiceden
+      contributors: springzh
+    config:
+      tags: ['magiceden','polygon','events']
+    description: >
+        Magic Eden events on Polygon
+    columns:
```
[the remark this answers is not in the document]
Comment (springzh) [reply models/magiceden/polygon/magiceden_polygon_schema.yml:13]: Added to unique_trade_id field.

**Comment by springzh** [comment]
> Hey @springzh, thanks for tackling these!
> 
> I think we should be able to include the royalty payments into the magic eden model.
> 
> https://polygonscan.com/tx/0xc55e0fee57f076d1944f1e9d22932c9e89e37debaee70572c56c86aca969c1af
> 
> The order of:
> 
> * first token transfer -> payment
> * second token transfer -> marketplace fee
> * third token transfer -> royalty payment is consistent between all trades
> 
> Please incorporate royalty and marketplace fees into this PR

I've refactored it to get raw amount from evt_Transfer for erc20 payment and from traces for native token payment.

**Review by 0xRobin — APPROVED** [review]
:heavy_check_mark: lgtm!

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**